File: backend/app/api/regions.py
"""行政区划 API（懒加载树形选择器）"""

from fastapi import APIRouter, Depends, Query, Request
from sqlalchemy.orm import Session
from typing import Optional

from app.core.database import get_db
from app.core.exceptions import LocalizedHTTPException
from app.core.i18n import request_locale
from app.core.security import get_current_admin_user
from app.models.administrative_region import AdministrativeRegion
from app.models.user import User
from app.schemas.region import RegionResponse, RegionDetailResponse, RegionAncestor
from app.services.catalog_display import region_display_name
from app.services.region_seed import upsert_administrative_regions, need_region_seed

router = APIRouter(tags=["行政区划"])


def _region_to_response(region: AdministrativeRegion, locale: str) -> RegionResponse:
    """ORM → Pydantic 响应"""
    return RegionResponse(
        id=region.id,
        code=region.code,
        name=region.name,
        display_name=region_display_name(region, locale),
        name_en=region.name_en,
        level=region.level,
        iso_country=region.iso_country,
        path=region.path,
        has_children=False,  # 由调用方覆写
    )


def _count_children(db: Session, parent_id: int) -> bool:
    """判断指定节点是否有活跃子节点。"""
    return (
        db.query(AdministrativeRegion.id)
        .filter(
            AdministrativeRegion.parent_id == parent_id,
            AdministrativeRegion.is_active == True,
        )
        .first()
        is not None
    )


@router.get("/regions", response_model=list[RegionResponse])
@router.get("/regions/", response_model=list[RegionResponse])
def list_regions(
    request: Request,
    parent_id: Optional[int] = Query(None),
    level: Optional[int] = Query(None),
    db: Session = Depends(get_db),
):
    """懒加载子节点列表。

    - ``parent_id`` 给定 → 返回该节点的活跃子节点（按 code 排序）
    - ``level`` 给定（无 parent_id）→ 返回该层级所有节点
    - 默认 → 返回 level=0（国家）
    """
    query = db.query(AdministrativeRegion).filter(
        AdministrativeRegion.is_active == True,
    )

    if parent_id is not None:
        query = query.filter(AdministrativeRegion.parent_id == parent_id)

    if level is not None:
        query = query.filter(AdministrativeRegion.level == level)

    # 默认（两者均未给定）：返回顶层，即 level=0
    if parent_id is None and level is None:
        query = query.filter(AdministrativeRegion.level == 0)

    locale = request_locale(request)
    rows = query.order_by(AdministrativeRegion.code).all()

    results = []
    for r in rows:
        resp = _region_to_response(r, locale)
        resp.has_children = _count_children(db, r.id)
        results.append(resp)
    return results


@router.get("/regions/{region_id}", response_model=RegionDetailResponse)
@router.get("/regions/{region_id}/", response_model=RegionDetailResponse)
def get_region(request: Request, region_id: int, db: Session = Depends(get_db)):
    """获取单个行政区划详情（含祖先链）。"""
    region = (
        db.query(AdministrativeRegion)
        .filter(
            AdministrativeRegion.id == region_id,
            AdministrativeRegion.is_active == True,
        )
        .first()
    )
    if not region:
        raise LocalizedHTTPException(status_code=404, message="行政区划不存在")

    locale = request_locale(request)

    # 通过 path 字段解析祖先链
    ancestors: list[RegionAncestor] = []
    if region.path:
        parts = region.path.split("/")
        for code in parts:
            if not code:
                continue
            ancestor = (
                db.query(AdministrativeRegion)
                .filter(
                    AdministrativeRegion.code == code,
                    AdministrativeRegion.is_active == True,
                )
                .first()
            )
            if ancestor and ancestor.id != region.id:
                ancestors.append(
                    RegionAncestor(
                        id=ancestor.id,
                        code=ancestor.code,
                        name=ancestor.name,
                        display_name=region_display_name(ancestor, locale),
                        level=ancestor.level,
                    )
                )

    resp = _region_to_response(region, locale)
    resp.has_children = _count_children(db, region.id)
    return RegionDetailResponse(
        **resp.model_dump(),
        ancestors=ancestors,
    )


@router.get("/admin/regions/seed-status")
def region_seed_status(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """行政区划数据状态：各级数量 + 是否需要 seed。仅管理员。"""
    counts = {}
    for lv in range(4):
        counts[str(lv)] = (
            db.query(AdministrativeRegion)
            .filter(
                AdministrativeRegion.level == lv,
                AdministrativeRegion.is_active == True,
            )
            .count()
        )
    return {
        "counts": counts,
        "needed": need_region_seed(db),
        "total": sum(counts.values()),
    }


@router.post("/admin/regions/seed")
def region_seed_run(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """手动 upsert 行政区划数据（更新/补缺失）。仅管理员。同步执行。"""
    result = upsert_administrative_regions(db)
    return result

File: frontend/src/api/local/handlers/regions.ts
// Administrative regions handler (lazy-load tree selector) - local mode
// Data stored in IndexedDB 'regions' store; seed from src/data/regions/*.json
import { getAll, getDb } from '../database.ts'
import { localError } from '../../../utils/localErrors.ts'
import i18n from '../../../plugins/i18n.ts'

interface RegionRow {
  id: number
  code: string
  name: string
  name_en?: string | null
  level: number
  parent_id?: number | null
  iso_country?: string | null
  path?: string | null
  is_active?: boolean
}

function isActive(r: any): boolean {
  return r.is_active !== false
}

export function regionDisplayName(
  row: Pick<RegionRow, 'code' | 'name' | 'name_en' | 'iso_country' | 'level'>,
): string {
  const locale = i18n.global.locale.value
  if (Number(row.level) !== 0) {
    return locale === 'zh-CN' ? row.name : row.name_en ?? row.name
  }

  const regionCode = row.iso_country ?? row.code
  return new Intl.DisplayNames([locale], { type: 'region' }).of(regionCode) ?? row.name
}

function buildParentSet(rows: any[]): Set<number> {
  const set = new Set<number>()
  for (const r of rows) {
    if (isActive(r) && r.parent_id != null) set.add(Number(r.parent_id))
  }
  return set
}

// GET /regions - lazy-load children by parent_id or level (default level=0)
export async function listRegions(
  _params: Record<string, string>,
  query?: any,
): Promise<any[]> {
  const all: any[] = await getAll('regions')
  const active = all.filter(isActive)

  let subset: any[]
  if (query?.parent_id != null && query.parent_id !== '') {
    const pid = Number(query.parent_id)
    subset = active.filter((r) => Number(r.parent_id) === pid)
  } else if (query?.level != null && query.level !== '') {
    const lv = Number(query.level)
    subset = active.filter((r) => Number(r.level) === lv)
  } else {
    subset = active.filter((r) => Number(r.level) === 0)
  }

  subset.sort((a, b) => String(a.code || '').localeCompare(String(b.code || '')))

  const parentSet = buildParentSet(active)
  return subset.map((r) => ({
    id: r.id,
    code: r.code,
    name: r.name,
    name_en: r.name_en ?? null,
    level: r.level,
    iso_country: r.iso_country ?? null,
    path: r.path ?? null,
    has_children: parentSet.has(Number(r.id)),
    display_name: regionDisplayName(r),
  }))
}

// GET /regions/:id - detail with ancestor chain parsed from path field
export async function getRegion(params: Record<string, string>): Promise<any> {
  const id = Number(params.id)
  const all: any[] = await getAll('regions')
  const region = all.find((r) => Number(r.id) === id && isActive(r))
  if (!region) throw localError('regionNotFound', 404, { id })

  const ancestors: any[] = []
  const pathCodes = String(region.path || '').split('/').filter(Boolean)
  for (const code of pathCodes) {
    if (code === region.code) continue
    const anc = all.find((r) => r.code === code && isActive(r))
    if (anc) {
      ancestors.push({
        id: anc.id,
        code: anc.code,
        name: anc.name,
        level: anc.level,
        display_name: regionDisplayName(anc),
      })
    }
  }

  const parentSet = buildParentSet(all.filter(isActive))
  return {
    id: region.id,
    code: region.code,
    name: region.name,
    name_en: region.name_en ?? null,
    level: region.level,
    iso_country: region.iso_country ?? null,
    path: region.path ?? null,
    has_children: parentSet.has(Number(region.id)),
    ancestors,
    display_name: regionDisplayName(region),
  }
}

// GET /admin/regions/seed-status
export async function seedStatus(): Promise<any> {
  const all: any[] = await getAll('regions')
  const active = all.filter(isActive)
  const counts: Record<string, number> = { '0': 0, '1': 0, '2': 0, '3': 0 }
  for (const r of active) {
    const lv = String(r.level)
    if (lv in counts) counts[lv]++
  }
  const cnProvinces = active.filter(
    (r) => Number(r.level) === 1 && r.iso_country === 'CN',
  ).length
  return {
    counts,
    needed: cnProvinces === 0,
    total: active.length,
  }
}

// POST /admin/regions/seed - upsert all region data from JSON files
// Uses dynamic import() instead of fetch() to bypass service worker interception
export async function seedRegions(): Promise<any> {
  const [countriesMod, chinaMod, namesMod] = await Promise.all([
    import('../../../data/regions/countries.json'),
    import('../../../data/regions/china_pca.json'),
    import('../../../data/regions/country_names_zh.json'),
  ])
  const countries = countriesMod.default
  const china = chinaMod.default
  const namesZh = namesMod.default

  // Read existing for upsert; determine starting id
  const existing: any[] = await getAll('regions')
  const codeMap = new Map<string, any>()
  let maxId = 0
  for (const r of existing) {
    if (r.code) codeMap.set(r.code, r)
    const n = Number(r.id)
    if (Number.isFinite(n) && n > maxId) maxId = n
  }
  let nextId = maxId + 1
  let created = 0
  let skipped = 0
  const records: any[] = []

  function upsert(
    code: string,
    name: string,
    level: number,
    parentId: number | null,
    isoCountry: string,
    path: string,
    nameEn?: string,
  ): number {
    const row = codeMap.get(code)
    if (row) {
      row.name = name
      row.level = level
      row.parent_id = parentId
      row.iso_country = isoCountry
      row.path = path
      if (nameEn !== undefined) row.name_en = nameEn
      row.is_active = true
      records.push(row)
      skipped++
      return row.id
    }
    const newRegion: RegionRow = {
      id: nextId++,
      code,
      name,
      name_en: nameEn ?? null,
      level,
      parent_id: parentId,
      iso_country: isoCountry,
      path,
      is_active: true,
    }
    codeMap.set(code, newRegion)
    records.push(newRegion)
    created++
    return newRegion.id
  }

  // Countries (level 0)
  for (const c of countries) {
    const alpha2 = c['alpha-2'] || ''
    const enName = c.name || ''
    if (!alpha2 || !enName) continue
    const zh = namesZh[alpha2] || enName
    upsert(alpha2, zh, 0, null, alpha2, alpha2, enName)
  }

  // Ensure China exists
  let cnId: number
  const cnRow = codeMap.get('CN')
  if (cnRow) {
    cnId = cnRow.id
  } else {
    cnId = upsert('CN', '\u4e2d\u56fd', 0, null, 'CN', 'CN', 'China')
  }

  // China provinces / cities / counties (level 1/2/3)
  for (const prov of china) {
    const pCode = prov.code
    const pName = prov.name
    const pId = upsert(pCode, pName, 1, cnId, 'CN', 'CN/' + pCode)
    for (const city of prov.children || []) {
      const cCode = city.code
      const cName = city.name
      const cId = upsert(cCode, cName, 2, pId, 'CN', 'CN/' + pCode + '/' + cCode)
      for (const county of city.children || []) {
        upsert(
          county.code,
          county.name,
          3,
          cId,
          'CN',
          'CN/' + pCode + '/' + cCode + '/' + county.code,
        )
      }
    }
  }

  // Bulk write in a single transaction
  const db = await getDb()
  const tx = db.transaction('regions', 'readwrite')
  // Fire all puts without awaiting each (IndexedDB processes them in order within the tx)
  for (const r of records) {
    tx.store.put(r)
  }
  await tx.done

  return { created, skipped, errors: 0 }
}
